src/components/TreeView/TreeView.tsx

```diff
@@ -0,0 +1,178 @@
+import React, { useState } from 'react';
+import {
+  DndContext,
+  DragOverlay,
+  DropAnimation,
+  MeasuringStrategy,
+  closestCenter,
+  defaultDropAnimation,
+} from '@dnd-kit/core';
+import {
+  SortableContext,
+  verticalListSortingStrategy,
+} from '@dnd-kit/sortable';
+import { CSS } from '@dnd-kit/utilities';
+
+import { getSubRowCount, setProperty } from './common/utils';
+import type { TreeViewProps, TreeViewRow } from './TreeView.types';
+import SortableTreeItem from './components/SortableTreeItem';
+import { Surface } from '../layout';
+import { useDnD } from './hooks/useDnD';
+
+const measuring = {
+  droppable: {
+    strategy: MeasuringStrategy.Always,
+  },
+};
+
+const dropAnimationConfig: DropAnimation = {
+  keyframes({ transform }) {
+    return [
+      { opacity: 1, transform: CSS.Transform.toString(transform.initial) },
+      {
+        opacity: 0,
+        transform: CSS.Transform.toString({
+          ...transform.final,
+          x: transform.final.x + 5,
+          y: transform.final.y + 5,
+        }),
+      },
+    ];
+  },
+  easing: 'ease-out',
+  sideEffects({ active }) {
+    active.node.animate([{ opacity: 0 }, { opacity: 1 }], {
+      duration: defaultDropAnimation.duration,
+      easing: defaultDropAnimation.easing,
+    });
+  },
+};
+
+function TreeView<D>({
+  data,
+  rowHeight = 56,
+  isCollapsible = true,
+  isSortable = true,
+  renderPrimaryContent,
+  renderSecondaryContent,
+  onDragCancel,
+  onDragEnd,
+  onDragMove,
+  onDragOver,
+  onDragStart,
+  onRowClick,
+  activeRowId,
+  onActiveRowIdChange,
+  rowActions,
+}: TreeViewProps<D>) {
+  const [items, setItems] = useState(() => data);
+
+  const {
+    activeId,
+    activeItem,
+    announcements,
+    flattenedItems,
+    projected,
+    sensors,
+    sortedIds,
+    handleDragCancel,
+    handleDragEnd,
+    handleDragMove,
+    handleDragOver,
+    handleDragStart,
+  } = useDnD({ items, setItems, onDragEnd });
+
+  const handleCollapse = (id: string) => {
+    setItems((prevItems) =>
+      setProperty(
+        prevItems,
+        id,
+        'collapsed',
+        (value) => !value as TreeViewRow<D>['collapsed'],
+      ),
+    );
+  };
+
+  return (
+    <Surface background="white" elevation={2} radius="md" hasBorder>
+      <DndContext
+        accessibility={{ announcements }}
+        collisionDetection={closestCenter}
+        measuring={measuring}
+        sensors={sensors}
+        onDragCancel={(e) => {
+          onDragCancel?.(e);
+          handleDragCancel();
+        }}
+        onDragEnd={(e) => {
+          handleDragEnd(e);
```

Comment:
Maybe missing a `onDragEnd?.(e);`?

Reply:
Not really the `onDragEnd` is called from within the `handleDragEvent` since we need to pass internal data from that function.
https://github.com/securityscorecard/design-system/pull/1015/files#diff-e3f7ff2a089182aab349911eef777a4b5c0f60961141593f0804069b7bbd339cR162